sql: fix cast from string to array with width

RaduBerinde opened the pull request:
The code that performs a cast from string to array doesn't take into account the
width of the string type it is producing. This doesn't show up when the input is
a literal constant because the constant is first parsed as an array of (vanilla)
strings which is then subjected to a cast (to array of strings with width).

Fixes #50132.

Release note (bug fix): Fixed some cases of casting a string to a width-limited
string array.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50153)
<!-- Reviewable:end -->

otan commented on pkg/sql/sem/tree/parse_string.go:
```diff
@@ -53,6 +53,12 @@ func ParseAndRequireString(t *types.T, s string, ctx ParseTimeContext) (Datum, e
+		// If the string type specifies a limit we truncate to that limit:
+		//   'hello'::CHAR(2) -> 'he'
+		// This is true of all the string type variants.
+		if t.Width() > 0 && int(t.Width()) < len(s) {
```
interesting that parsing and storing strings that are too long have different results.

```
otan=# create temp table a (a varchar(1));
CREATE TABLE

otan=# insert into a values ('123');
ERROR:  value too long for type character varying(1)
```

Should this be rune count like we do here: https://github.com/cockroachdb/cockroach/blob/0f1a1631ea298c9818e4cf8e760a063cd17a9c63/pkg/sql/sqlbase/column_type_properties.go#L54 ?

RaduBerinde replied:
Hmm. I copied this from PerformCast (string->string case). I think you're right though, postgres docs say that the limit is n "characters (not bytes)".

otan reviewed (commented):
wondering if we should backport this in case people relied on this behaviour in older versions (i hope not?!).

RaduBerinde commented:
I will open a separate PR to fix the unicode stuff. And yeah, we should backport both (I already added the backport labels).

otan reviewed (approved):
ok! lgtm!

RaduBerinde commented:
Opened #50156. I will rebase this on top of that.

RaduBerinde committed:
sql: fix cast from string to array with width

The code that performs a cast from string to array doesn't take into
account the width of the string type it is producing. This doesn't
show up when the input is a literal constant because the constant is
first parsed as an array of (vanilla) strings which is then subjected
to a cast (to array of strings with width).

Fixes #50132.

Release note (bug fix): Fixed some cases where casting a string to a
width-limited string array was not truncating the string.

RaduBerinde commented:
bors r+
